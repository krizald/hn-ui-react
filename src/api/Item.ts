export default interface Item {
  Id: number;
  deleted: boolean;
  type: string;
  by: string;
  time: Date;
}
