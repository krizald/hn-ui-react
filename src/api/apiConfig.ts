export default {
  databaseURL: 'https://hacker-news.firebaseio.com',
  version: '/v0',
};
