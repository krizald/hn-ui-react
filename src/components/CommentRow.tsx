import { Grid, makeStyles, Table, TableCell, TableRow } from '@material-ui/core';
import he from 'he';
import moment from 'moment';
import React, { FC } from 'react';
import Constants from '../constants';
import { ItemModel } from '../models';

interface CommentRowProps {
  data: ItemModel;
  incident: number;
}

const useStyles = makeStyles({
  bold: { fontWeight: 'bold' },
});

const CommentRow: FC<CommentRowProps> = (props: CommentRowProps) => {
  const classes = useStyles();
  const { data, incident } = props;
  return (
    <>
      <TableRow key={data.id}>
        <Table>
          <TableRow key={data.id}>
            <TableCell style={{ width: `${incident.toString()}px` }}>
              <span />
            </TableCell>
            <TableCell component="th" scope="row">
              <Grid container spacing={3}>
                <Grid item xs={3}>
                  <span className={classes.bold}>Author</span>
                </Grid>
                <Grid item xs={3}>
                  <span data-testid="comment-author">{data.by}</span>
                </Grid>
                <Grid item xs={3}>
                  <span className={classes.bold}>date time</span>
                </Grid>
                <Grid item xs={3}>
                  <span>
                    {moment(data.createDate).format(Constants.dateTimeFormat)}
                  </span>
                </Grid>
                <Grid item xs={12} data-testid="comment-text">
                  {React.createElement('div', {
                    dangerouslySetInnerHTML: {
                      __html: he.decode(data.text ? data.text : ''),
                    },
                  })}
                </Grid>
              </Grid>
            </TableCell>
          </TableRow>
        </Table>
      </TableRow>
      {data.kidsItems && data.kidsItems.length > 0
        ? data.kidsItems.map((row: ItemModel) => {
            return (
              <CommentRow
                key={row.id}
                data={row}
                incident={incident + Constants.incidentPerStop}
              />
            );
          })
        : null}
    </>
  );
};
export default CommentRow;
