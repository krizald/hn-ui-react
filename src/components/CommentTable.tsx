import React, { FC } from 'react';
import { TableContainer, Table } from '@material-ui/core';
import { ItemModel } from '../models';
import { CommentRow } from '.';

interface CommentTableProps {
  data: ItemModel[];
  incident: number;
}

const CommentTable: FC<CommentTableProps> = (props: CommentTableProps) => {
  const { data, incident } = props;
  return (
    <TableContainer>
      <Table>
        {data.map((row: ItemModel) => (
          <CommentRow key={row.id} data={row} incident={incident} />
        ))}
      </Table>
    </TableContainer>
  );
};

export default CommentTable;
