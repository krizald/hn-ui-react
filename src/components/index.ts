export { default as LinkCellRenderer } from './LinkCellRenderer';
export { default as ItemGrid } from './ItemGrid';
export { default as ItemCardRenderer } from './ItemCardRenderer';
export { default as CommentTable } from './CommentTable';
export { default as CommentRow } from './CommentRow';
