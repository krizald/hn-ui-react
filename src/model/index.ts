export * from './Item';
export { default as ItemModel } from './Item';
