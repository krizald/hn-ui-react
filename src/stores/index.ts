export { default as NumericDictionary } from './numericDictionary';
export { default as StoryStore } from './storyStore';
