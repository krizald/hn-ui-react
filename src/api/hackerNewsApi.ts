import axios, { AxiosResponse } from 'axios';
import apiConfig from './apiConfig';
import Item from '../model';

export default class HackerNewsApi {
  fetchItem = (id: number): Promise<AxiosResponse<Item>> => {
    const query = `${apiConfig.databaseURL + apiConfig.version}/item/${id}.json`;
    return axios.get(query);
  };
}
