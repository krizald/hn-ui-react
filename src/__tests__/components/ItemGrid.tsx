import { mount } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { ItemGrid } from '../../components';
import { HackerNewsApi } from '../../api';
import { StoryStore } from '../../stores';
import { ItemModel } from '../../model';

jest.mock('../../api');

describe('ItemGrid test suite', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('should render, no api call', () => {
    HackerNewsApi.prototype.fetchItems = jest.fn();
    act(() => {
      mount(<ItemGrid itemId={[]} />);
    });
    expect(HackerNewsApi.prototype.fetchItems).toHaveBeenCalledTimes(0);
  });

  test('should render, api called', () => {
    HackerNewsApi.prototype.fetchItems = jest.fn();
    act(() => {
      mount(<ItemGrid itemId={[1, 2, 3, 4, 5]} />);
    });
    expect(HackerNewsApi.prototype.fetchItems).toHaveBeenCalledTimes(1);
  });

  test('should render, store api called', () => {
    HackerNewsApi.prototype.fetchItems = jest.fn();
    StoryStore.GetInstance().FetchItems = jest
      .fn()
      .mockImplementationOnce(() =>
        Promise.resolve([new ItemModel({ id: 1 })] as ItemModel[]),
      );
    const wrapper = mount(<ItemGrid itemId={[1, 2, 3, 4, 5]} />);
    act(() => {
      wrapper.update();
    });
    expect(HackerNewsApi.prototype.fetchItems).toHaveBeenCalledTimes(0);
    expect(StoryStore.GetInstance().FetchItems).toHaveBeenCalledTimes(1);
    expect(StoryStore.GetInstance().FetchItems).toHaveBeenCalledWith([
      1,
      2,
      3,
      4,
      5,
    ]);
    // const aggrid = wrapper.find('AgGridReact');
    // const rowData = aggrid.prop('rowData') as ItemModel[];
    // expect(rowData).toBe([]);
  });
});
