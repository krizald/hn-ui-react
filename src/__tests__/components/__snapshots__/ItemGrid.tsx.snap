// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemGrid test suite Render: should render multiple items 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "height:100%",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`ItemGrid test suite Render: should render nothing 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "height:100%",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`ItemGrid test suite Render: should render single item 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "height:100%",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
